Document Security Advisory process for maintainers (#3556)
openenclave/openenclave: docs/Community/SecurityAdvisoryHandling.md

# Security Advisory Process for Maintainers

## Reporting

For instructions on reporting a vulnerability, refer to the Open Enclave SDK [SECURITY.md](https://github.com/openenclave/openenclave/security/policy#reporting-a-vulnerability).

Vulnerability reporting for the OE SDK is currently still being handled through the [Microsoft Security Response Center (MSRC)](https://www.microsoft.com/en-us/msrc) as they:

- provide a 24-hour turnaround time for acknowledging reports
- maintain a [PGP key](https://www.microsoft.com/en-us/msrc/pgp-key-msrc#:~:text=MSRC%20PGP%20Key%20Last%20updated%3A%20October%2031%2C%202019,us.%20Click%20here%20to%20submit%20a%20security%20vulnerability) for secure reporting of vulnerabilities to secure@microsoft.com.

Once a vulnerability related to the Open Enclave SDK is reported to MSRC, the report is forwarded to the committers responsible for reviewing security issues:

- Anand Krishnamoorthi (@anakrish)

> - Ming-wei Shih (@mingweishih)

Reviewer: ```suggestion
- Ming-Wei Shih (@mingweishih)
```

- Simon Leet (@codemonkeyleet)

> - [ ] TODO: Formalize SIG-Security and its membership with the CGC.

## Handling Embargoed Vulnerabilities

The security reviewers are responsible for the initial triage of the report. This includes:

- reaching out to the appropriate OE SDK area owners or subject matter experts

> - evaluating the severity of the issue against the bug bar

Reviewer 2: evaluating the severity and exploitability of the issue

TODO: OE should adopt baseline guidance of how OE vulnerabilities might map to Critical, Important, Moderate, and Low severity.

Author: This is the same TODO as indicated on line 29.

> - writing up the initial assessment and plan of action for MSRC

Reviewer: ```suggestion
- writing up an initial assessment and a plan of action for MSRC
```

- coordinating with MSRC on additional information, acknowledgment, and public disclosure plans with the reporters of the issue

> - [ ] TODO: Clean up and finalize the security bug triage bar (PR #2634).

The report is treated as under embargo and will only be discussed using private communication channels.
All individuals entrusted with the information should exercise due diligence in maintaining its confidentiality.

> - [ ] TODO: Formalize the protected channel of communication for discussing the development of embargoed bug fixes (e.g., Protected mailing list, Element private channel).

> If OE SDK needs to issue a Security Advisory to address the reported concern, one of the security reviewers will:

Reviewer 2: It's not clear that this should be done on a private branch.

Author: Is this a comment to Dave's comment? I not sure how this comment relates here. See the update with the link for clarification on the exact definition of the temporary private fork for embargoed security fixes.

1. Create a new GitHub [Security Advisory (SA)](https://github.com/openenclave/openenclave/security/advisories) for OE SDK.
   1. The description of the SA should accurately capture the nature, severity, and workarounds for the issue.
      1. It does not need to capture the full technical details of the vulnerability or fix for brevity.
2. Request a Common Vulnerability Enumeration (CVE) through GitHub.
   1. This was previously done through MSRC as a CVE Numbering Authority (CNA), but OE has shifted to using GitHub as the CNA as a Linux Foundation project.
3. Assign the appropriate collaborators to the SA, which includes:
   1. The assignee(s) responsible for fixing the issue.
   2. Any area owners or subject matter experts needed for review.
   3. CI/CD coordinators for private testing and integration with the release. (e.g. @brmclaren).

> 4. Create a temporary private fork to implement the code changes as part of the SA.

Reviewer 3: Clarifying question: should the private fork be an individual's fork?  (this is how I read it, so just confirming)

Author: No, this is special temporary private fork off the OE SDK project created through GitHub SA, not an individual's private fork. I've added a link to the creation process to clarify.

Reviewer 3: Looks good thanks.

   1. The assignee should follow the GitHub instructions and create an `advisory-fix` branch for their code changes.
      1. The code changes should include an update to [CHANGELOG.md](/CHANGELOG.md) under the `### Security` subheader describing the issue addressed.
   2. Once the changes are ready for review, they should be submitted as a PR to the master branch of the temporary fork.

>       1. GitHub does not currently appear to support protecting the master branch in temporary forks, so maintainers will need to enforce this process manually.

Reviewer: Consider clarifying "protecting the master branch in temporary forks" a bit more. Does it mean any changes directly made to the master branch will be public?

Author: I've clarified the wording around branch protection as requiring PR reviews for merging.

5. Reach out to the CI/CD maintainers to establish a private CI/CD pipeline for testing code changes.

> By default, all SA are targeted for inclusion into the next release of the OE SDK, which usually happens quarterly.

Reviewer 3: ```suggestion
By default, all SAs are targeted for inclusion into the next release of the OE SDK, which usually happens quarterly.
```

If the fix cannot be implemented or verified in time for an upcoming release, it can be included in a following patch version release.

> The security reviewers are responsible for deciding if a patch release is needed to address vulnerability before its scheduled public disclosure date.

Reviewer 3: ```suggestion
The security reviewers are responsible for deciding if a patch release is needed to address a vulnerability before its scheduled public disclosure date.
```

Once the security fix is reviewed, tested, and merged into the temporary fork master branch for release,
the assignee and the release manager will coordinate integrating the fix into the release branch.
This usually includes:

1. Rebasing the temporary fork master branch to the project's master branch and testing the merged branch in private CI/CD.
2. Publishing the SA and merging its master branch into the project master branch.

>    1. At this point, the vulnerability is considered publicly disclosed, and a new version of the OE SDK packages with the fix should be made available as quickly as possible.

Reviewer 2: OE should not "publish the SA" before the OE SDK packages are available. Is it feasible to produce OE SDK packages from the temporary fork master such that "publish SA", merge, and package availability are simultaneous?

Author: Possible, but not feasible. We have tried to do this in the past and the manual coordination and costs involved creates so many additional points of failure that it's not worth it. It requires building a package off a private fork, publishing it, _then_ merging the change into public master, cherry picking into the release branch, and verifying that the published package is what the public repo would have produced. Because public master is always merging changes, there's the need to resolve merge conflicts, and that invalidates the exact match with the release package from the private. Using the GitHub provided tooling and workflow simplifies this considerably.

In discussion with MSRC, they were fine with the delta being about half a day when we were collaborating with them on the publication of CVEs to update the link on the updated packages.

Author: @markg-intel That's an interesting point, and one that we've not had to specifically deal with before. Our communications with the researchers have been mediated by MSRC and we've not gotten feedback to this effect, but is there a way you would recommend amending the text to accommodate this possibility?

> 3. Cherry-picking the fix to the release branch.

Reviewer 2: Perhaps: "Selectively applying the fix to the release branch."

Author: No, cherry-picking is a term specific to Git and is the exact meaning intended here to preserve parity in the the commit history.

4. Building and releasing the OE SDK packages with the fix.

> 5. Updating the published SA with a pointer to the released packages containing the fix.

Reviewer 2: Please see early comments about avoiding delay between publishing the SA and providing released packages containing the fix.

## Disclosure

> Open Enclave SDK project currently relies on the principles for [Coordinated Vulnerability Disclosure (CVD)](https://www.microsoft.com/en-us/msrc/cvd) put forth by MSRC.

Reviewer 3: ```suggestion
The Open Enclave SDK project currently relies on the principles for [Coordinated Vulnerability Disclosure (CVD)](https://www.microsoft.com/en-us/msrc/cvd) put forth by MSRC.
```

- writing up an initial assessment and a plan of action for MSRC
- coordinating with MSRC on additional information, acknowledgment, and public disclosure plans with the reporters of the issue

> - [ ] TODO: Clean up and finalize the security bug triage bar (PR #2634).

The report is treated as under embargo and will only be discussed using private communication channels.
All individuals entrusted with the information should exercise due diligence in maintaining its confidentiality.

> - [ ] TODO: Formalize the protected channel of communication for discussing the development of embargoed bug fixes (e.g., Protected mailing list, Element private channel).

If OE SDK needs to issue a Security Advisory to address the reported concern, one of the security reviewers will:

1. [Create](https://docs.github.com/en/github/managing-security-vulnerabilities/creating-a-security-advisory#creating-a-security-advisory)
   a new GitHub [Security Advisory (SA) for OE SDK](https://github.com/openenclave/openenclave/security/advisories).
   1. The description of the SA should accurately capture the nature, severity, and workarounds for the issue.

>       1. It does not need to capture the full technical details of the vulnerability or fix for brevity.

Author: I'll submit a follow-up PR to amend the writing here.

2. [Request a Common Vulnerability Enumeration (CVE)](https://docs.github.com/en/github/managing-security-vulnerabilities/publishing-a-security-advisory#requesting-a-cve-identification-number) through GitHub.
   1. This was previously done through MSRC as a CVE Numbering Authority (CNA), but OE has shifted to using GitHub as the CNA as a Linux Foundation project.
3. [Assign the appropriate collaborators](https://docs.github.com/en/github/managing-security-vulnerabilities/adding-a-collaborator-to-a-security-advisory) to the SA, which includes:
   1. The assignee(s) responsible for fixing the issue.
   2. Any area owners or subject matter experts needed for review.
   3. CI/CD coordinators for private testing and integration with the release. (e.g., [@brmclaren](https://github.com/brmclaren)).
4. [Create a temporary private fork](https://docs.github.com/en/github/managing-security-vulnerabilities/collaborating-in-a-temporary-private-fork-to-resolve-a-security-vulnerability#creating-a-temporary-private-fork)
   to implement the code changes as part of the SA.
   1. The assignee should follow the GitHub instructions and create an `advisory-fix` branch for their code changes.

>       1. The code changes should include an update to [CHANGELOG.md](/CHANGELOG.md) under the `### Security` subheader describing the issue addressed.

Author: I'll amend this to specify that it should include a link to the SA (which would contain information to the CVE as well)

   2. Once the changes are ready for review, they should be submitted as a PR to the master branch of the temporary fork.
      1. GitHub does not currently appear to support requiring PR reviews to merge into the master branch of a temporary fork, so maintainers will need to enforce this process manually.
5. Reach out to the CI/CD maintainers to establish a private CI/CD pipeline for testing code changes.

By default, all SAs are targeted for inclusion into the next release of the OE SDK, which usually happens quarterly.
If the fix cannot be implemented or verified in time for an upcoming release, it can be included in a following patch version release.

> The security reviewers are responsible for deciding if a patch release is needed to address a vulnerability before its scheduled public disclosure date.

Author: @markg-intel I agree that timelines are usually influenced by the severity evaluation, I couldn't think of any specific procedural recommendation I could articulate based on that. Did you have any text you would like to suggest for the document to that effect?

Once the security fix is reviewed, tested, and merged into the temporary fork master branch for release,
the assignee and the release manager will coordinate integrating the fix into the release branch.
This usually includes:

1. Rebasing the temporary fork master branch to the project's master branch and testing the merged branch in private CI/CD.
2. Publishing the SA and merging its master branch into the project master branch.
   1. At this point, the vulnerability is considered publicly disclosed, and a new version of the OE SDK packages with the fix should be made available as quickly as possible.
3. Cherry-picking the fix to the release branch.
4. Building and releasing the OE SDK packages with the fix.

> 5. Updating the published SA with a pointer to the released packages containing the fix.

Reviewer 4: Can we know the "pointer" in advance such that SA could include sooner and basically avoid this step?

## Disclosure

The Open Enclave SDK project currently relies on the principles for [Coordinated Vulnerability Disclosure (CVD)](https://www.microsoft.com/en-us/msrc/cvd) put forth by MSRC.
The CVD includes a couple of policies such as:

- acknowledgment of vulnerability reports within 24 hours (via MSRC)

> - supporting public disclosure within 90 days as the baseline, and negotiating that with vulnerability reporters and necessary

Reviewer 4: incomplete sentence here. necessary "trusted stakeholders"?

Author: Not sure how this line got truncated; it's supposed to say "necessary parties" which would include engineering teams and vendors as discussed in the CVD. 

The "trusted stakeholder" term is only introduced after this to more specifically discuss a group OE as an open source project needs to define for its governance.
